# 高校数学とPython Day 2【Python版】

- 城北中学校・高等学校　中学3年・高校1年
- 夏期講習会III 2026/8/24~2026/8/28
- 担当:清水団

## 本日のテーマ:整数問題をプログラムで考えよう!

Julia版と同じ内容を **Python** で体験するバージョンです。言語ごとの書き方の違いにも注目してみてください。

### 5日間の学習予定

- **Day 1**:基本計算・for文・if文 ✅
- **Day 2**:整数問題をプログラムで考えよう ← 今日
- **Day 3**:場合の数・組合せを実験しよう
- **Day 4**:関数・グラフ・最大最小
- **Day 5**:確率・シミュレーション


## 約数を全部求めよう

In [1]:
def divisors(n):
    return [d for d in range(1, n + 1) if n % d == 0]

divisors(24)

[1, 2, 3, 4, 6, 8, 12, 24]

In [2]:
len(divisors(24)), sum(divisors(24))  # 約数の個数と和

(8, 60)

## 素数判定と sympy

Colabには数学パッケージ **sympy** が最初から入っています。

In [3]:
# 自作:約数がちょうど2個なら素数
def isprime_mine(n):
    return n >= 2 and len(divisors(n)) == 2

isprime_mine(17), isprime_mine(21)

(True, False)

In [4]:
from sympy import isprime, primerange, factorint

isprime(9999999967)

True

In [5]:
list(primerange(1, 100))  # 100以下の素数

[2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97]

In [6]:
factorint(2026)  # 素因数分解 {素数: 指数}

{2: 1, 1013: 1}

## エラトステネスの篩(ふるい)

In [7]:
def eratosthenes(n):
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    for p in range(2, n + 1):
        if is_prime[p]:
            for q in range(2*p, n + 1, p):
                is_prime[q] = False
    return [p for p in range(n + 1) if is_prime[p]]

eratosthenes(100)

[2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97]

In [8]:
# sympyの結果と一致するか確認
eratosthenes(1000) == list(primerange(1, 1001))

True

In [9]:
# 素数の割合の実験
for n in [100, 1000, 10000, 100000]:
    k = len(list(primerange(1, n + 1)))
    print(f"{n} 以下の素数: {k} 個(割合 {k/n*100:.2f}%)")

100 以下の素数: 25 個(割合 25.00%)
1000 以下の素数: 168 個(割合 16.80%)
10000 以下の素数: 1229 個(割合 12.29%)
100000 以下の素数: 9592 個(割合 9.59%)


## 最大公約数とユークリッドの互除法

In [10]:
import math
math.gcd(1071, 1029)  # 標準装備

21

In [11]:
def my_gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

my_gcd(1071, 1029)

21

In [12]:
import math
math.lcm(12, 18), math.gcd(12, 18) * math.lcm(12, 18) == 12 * 18

(36, True)

## 完全数を探そう

In [13]:
def proper_sum(n):
    return sum(divisors(n)) - n

for n in range(2, 10001):
    if proper_sum(n) == n:
        print(f"{n} は完全数! 約数: {divisors(n)}")

6 は完全数! 約数: [1, 2, 3, 6]
28 は完全数! 約数: [1, 2, 4, 7, 14, 28]
496 は完全数! 約数: [1, 2, 4, 8, 16, 31, 62, 124, 248, 496]
8128 は完全数! 約数: [1, 2, 4, 8, 16, 32, 64, 127, 254, 508, 1016, 2032, 4064, 8128]


## 双子素数を探そう

In [14]:
for n in range(2, 101):
    if isprime(n) and isprime(n + 2):
        print(f"({n}, {n+2})")

(3, 5)
(5, 7)
(11, 13)
(17, 19)
(29, 31)
(41, 43)
(59, 61)
(71, 73)


## フェルマー数

Pythonの整数は**最初から何桁でもOK**(Juliaの big() は不要)。

In [15]:
for n in range(6):
    F = 2**(2**n) + 1
    print(f"F_{n} = {F} → 素数? {isprime(F)}")

F_0 = 3 → 素数? True
F_1 = 5 → 素数? True
F_2 = 17 → 素数? True
F_3 = 257 → 素数? True
F_4 = 65537 → 素数? True
F_5 = 4294967297 → 素数? False


In [16]:
factorint(2**32 + 1)  # F₅の素因数分解(オイラーの発見を1秒で!)

{641: 1, 6700417: 1}

## ピタゴラス数を探そう

In [17]:
for c in range(1, 51):
    for b in range(1, c + 1):
        for a in range(1, b + 1):
            if a*a + b*b == c*c:
                print(f"({a}, {b}, {c})")

(3, 4, 5)
(6, 8, 10)
(5, 12, 13)
(9, 12, 15)
(8, 15, 17)
(12, 16, 20)
(15, 20, 25)
(7, 24, 25)
(10, 24, 26)
(20, 21, 29)
(18, 24, 30)
(16, 30, 34)
(21, 28, 35)
(12, 35, 37)
(15, 36, 39)
(24, 32, 40)
(9, 40, 41)
(27, 36, 45)
(30, 40, 50)
(14, 48, 50)


## コラッツ予想:世界一有名な未解決問題

In [18]:
def collatz_steps(n):
    steps = 0
    while n != 1:
        n = n // 2 if n % 2 == 0 else 3*n + 1
        steps += 1
    return steps

# 1〜30のステップ数
for n in range(1, 31):
    print(f"{n} → {collatz_steps(n)} ステップ")

1 → 0 ステップ
2 → 1 ステップ
3 → 7 ステップ
4 → 2 ステップ
5 → 5 ステップ
6 → 8 ステップ
7 → 16 ステップ
8 → 3 ステップ
9 → 19 ステップ
10 → 6 ステップ
11 → 14 ステップ
12 → 9 ステップ
13 → 9 ステップ
14 → 17 ステップ
15 → 17 ステップ
16 → 4 ステップ
17 → 12 ステップ
18 → 20 ステップ
19 → 20 ステップ
20 → 7 ステップ
21 → 7 ステップ
22 → 15 ステップ
23 → 15 ステップ
24 → 10 ステップ
25 → 23 ステップ
26 → 10 ステップ
27 → 111 ステップ
28 → 18 ステップ
29 → 18 ステップ
30 → 18 ステップ


In [19]:
# 1〜10000で一番長い旅をする数は?
n_max = max(range(1, 10001), key=collatz_steps)
print(f"{n_max} が最長で {collatz_steps(n_max)} ステップ!")

6171 が最長で 261 ステップ!


## Day 2 の演習問題

### 問題1: 496が完全数であることを確かめよう
### 問題2: 2桁の素数の個数/200以下の双子素数
### 問題3: 好きな数でコラッツの旅、1〜100で最長の数
### 問題4: カプレカ数(6174)の実験

## 解答欄

In [20]:
# 問題1



In [21]:
# 問題2



In [22]:
# 問題3



In [23]:
# 問題4



## Day 2 まとめ

- リスト内包表記 `[d for d in ... if ...]` はJuliaとほぼ同じ
- 素数は sympy(isprime・primerange・factorint)
- Pythonの整数は最初から任意精度 —— フェルマー数も余裕
- 未解決問題(コラッツ予想)も数行で実験できるのはPythonも同じ!